### PR TITLE
Refactor repos custom list

### PR DESCRIPTION
**Pull request by wstephenson**

### CODE REVIEW

**Comment by suse-tests-pass** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          5 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>lib/rmt/cli/products.rb</td>
      <td>37</td>
      <td>Style/AndOr: Use <code>||</code> instead of <code>or</code>. (https://github.com/bbatsov/ruby-style-guide#no-and-or-or)</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>spec/lib/rmt/cli/repos_custom_spec.rb</td>
      <td>349</td>
      <td>Layout/IndentationWidth: Use 2 (not 0) spaces for indentation. (https://github.com/bbatsov/ruby-style-guide#spaces-indentation)</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>spec/lib/rmt/cli/repos_custom_spec.rb</td>
      <td>355</td>
      <td>Layout/BlockAlignment: <code>end</code> at 355, 10 is not aligned with <code>let(:rows) do</code> at 348, 6.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>spec/lib/rmt/cli/repos_custom_spec.rb</td>
      <td>373</td>
      <td>RSpec/NestedGroups: Maximum example group nesting exceeded [4/3]. (http://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/NestedGroups)</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>spec/lib/rmt/cli/repos_custom_spec.rb</td>
      <td>373</td>
      <td>Style/StringLiterals: Prefer single-quoted strings when you don’t need string interpolation or special symbols. (https://github.com/bbatsov/ruby-style-guide#consistent-string-literals)</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Unless this is a trivial change, please include a CHANGELOG entry.<br />
Run <code>osc vc</code> in the <code>package</code> directory to add one.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: lib/rmt/cli/products.rb [hunk 1]
```diff
@@ -1,29 +1,40 @@
+require 'csv'
+require 'terminal-table'
```
Comment (ikapelyukhin) [inline lib/rmt/cli/products.rb]: Move both requires to `array_printable.rb`?

File: spec/lib/rmt/cli/products_spec.rb [hunk 2]
```diff
@@ -1,4 +1,5 @@
 require 'rails_helper'
+require 'csv'
```
Comment (ikapelyukhin) [inline spec/lib/rmt/cli/products_spec.rb]: Should be just in `array_printable.rb`.

File: spec/lib/rmt/cli/repos_spec.rb [hunk 3]
```diff
@@ -1,4 +1,5 @@
 require 'rails_helper'
+require 'csv'
```
Comment (ikapelyukhin) [inline spec/lib/rmt/cli/repos_spec.rb]: Should be just in `array_printable.rb`.

File: lib/rmt/cli/array_printable.rb [hunk 4]
```diff
@@ -1,12 +1,25 @@
...
+module RMT::CLI::ArrayPrintable
 
-    array.all.each do |element|
-      rows << options.keys.map { |k| element[k] }
+  def format_array(array, array_options)
+    rows = array.map do |element|
+      array_options.keys.map { |k| element.public_send(k) }
     end
-    Terminal::Table.new headings: options.values, rows: rows
+    if options.csv
```
Comment (thutterer) [inline lib/rmt/cli/array_printable.rb]: I don't like that this module method has a hidden dependency on the Thor gem's options object. This should be part of the method definition.

File: package/rmt-server.changes [hunk 5]
```diff
@@ -1,3 +1,11 @@
+-------------------------------------------------------------------
+Wed May  9 09:36:02 UTC 2018 - wstephenson@suse.com
```
Comment (thutterer) [inline package/rmt-server.changes]: Not sure if you want to update that date.

File: lib/rmt/cli/array_printable.rb (hunk 4 above)
Comment (lagartoflojo) [reply lib/rmt/cli/array_printable.rb]: Agreed.

File: lib/rmt/cli/products.rb [hunk 6]
```diff
@@ -2,28 +2,36 @@
...
+        version: 'Version',
+        arch: 'Architecture',
+        product_string: 'Product string',
+        release_stage: 'Release stage',
+        'mirror?' => 'Mirror?',
+        last_mirrored_at: 'Last mirrored'
+      })
     end
-    puts 'Only enabled products are shown by default. Use the `--all` option to see all products.' unless options.all
+    puts 'Only enabled products are shown by default. Use the `--all` option to see all products.' unless options.all || options.csv
```
Comment (lagartoflojo) [inline lib/rmt/cli/products.rb]: `unless` with `||` => *mind blown*

Comment (lagartoflojo) [reply lib/rmt/cli/products.rb]: This new boolean condition needs a test.

File: lib/rmt/cli/products.rb [hunk 7]
```diff
@@ -2,28 +2,36 @@
...
+      puts format_array(products, {
+        id: 'ID',
+        name: 'Name',
+        version: 'Version',
+        arch: 'Architecture',
+        product_string: 'Product string',
+        release_stage: 'Release stage',
+        'mirror?' => 'Mirror?',
+        last_mirrored_at: 'Last mirrored'
+      })
```
Comment (lagartoflojo) [inline lib/rmt/cli/products.rb]: I like the explicitness of this, The previous `public_send` thing was not nice. :+1:

File: package/rmt-server.changes (hunk 5 above)
Comment (lagartoflojo) [reply package/rmt-server.changes]: not critical imo

**Comment by lagartoflojo** [comment]
Functionally, it works great. I also ran the support config module and it also works.

**Comment by wstephenson** [comment]
Refactored the module method and covered the --all mention as requested.  There's a lot of DRY possible in products_spec, but I am loathe to do it this week for obvious reasons.

**Pull request merged**